onecodex/analyses.py

```diff
@@ -536,6 +490,63 @@ def _get_highest_unclassified_tax_id(self, tax_id, level):
 
         return highest_tax_id
 
+    @staticmethod
+    def _make_labels_by_item_id(metadata, label):
+        """Make labels from
+
+        Parameters
+        ----------
+        metadata : `pandas.DataFrame`
+        label : `str` or `callable`
+
+        Returns
+        -------
+        `dict`
+            Keys are from metadata.index. Values are generated labels.
+        """
+        import pandas as pd
+
+        raw_result = {}
+
+        if isinstance(label, str):
+            if label in metadata.columns:
+                raw_result = dict(metadata[label].items())
+            else:
+                raise OneCodexException(
+                    "Label field {} not found. Choose from: {}".format(
+                        label, ", ".join(metadata.keys())
+                    )
+                )
+        elif callable(label):
+            for item_id, item_meta in metadata.to_dict(orient="index").items():
+                item_label = label(item_meta)
+                if not isinstance(item_label, six.string_types):
+                    wrong_type = type(item_label).__name__
+                    raise OneCodexException(
+                        "Expected string from label function, got: {}".format(wrong_type)
+                    )
+                raw_result[item_id] = item_label
+
+        elif label is not None:
+            wrong_type = type(label).__name__
+            raise OneCodexException(
+                "Expected string or callable for label, got: {}".format(wrong_type)
+            )
+
+        # add an incremented number to duplicate labels (e.g., same filename)
+        duplicates_counter = Counter(raw_result.values())
```

Comment:
Is the counter really faster than pandas `duplicated` calls? Also, should we drop `na` values first?

Reply:
Python dict speed would become and issue with 10K items. So I do not worry about the speed. `na` values are valid comment. I will see maybe I can convert dict to `pd.Series` ealier.

Reply:
Turns out dropping `na` was just an artifact of calling `where` and `duplicated`.